File: scenarios/flow-verifier/api/src/features/conflicts/flow/transition.ts
import type { ConflictResolutionState, ConflictResolutionEvent } from "./generated/runtime";
import { nextConflictResolutionStatus } from "./generated/runtime";

export const transitionConflictResolution = (state: ConflictResolutionState, event: ConflictResolutionEvent): ConflictResolutionState => {
  const next = nextConflictResolutionStatus(state.status, event.type);
  return { status: next } as ConflictResolutionState;
};

File: scenarios/treasury/api/internal/approval/flow/transition.go
package flow

import (
	"fmt"

	"treasury/internal/approval/flow/generated"
)

func TransitionApproval(state ApprovalState, event ApprovalEvent) (ApprovalState, error) {
	if err := CheckInvariants(state); err != nil {
		return state, err
	}
	next, err := generated.TransitionApprovalStatus(state.Status, event)
	return ApprovalState{Status: next}, err
}

func CheckInvariants(state ApprovalState) error {
	switch state.Status {
	case ApprovalQueued, ApprovalApproved, ApprovalDeclined, ApprovalExpired:
		return nil
	default:
		return fmt.Errorf("unknown approval status %q", state.Status)
	}
}
